internal/signalfx-agent/pkg/monitors/prometheusexporter/conversion.go

```diff
@@ -24,66 +22,110 @@ func counterExtractor(m *dto.Metric) float64 {
 	return m.GetCounter().GetValue()
 }
 
-func convertMetricFamily(mf *dto.MetricFamily) []*datapoint.Datapoint {
+func convertMetricFamily(sm pmetric.ScopeMetrics, mf *dto.MetricFamily, applyDimensions func(attrs pcommon.Map)) {
 	if mf.Type == nil || mf.Name == nil {
-		return nil
+		return
 	}
 	switch *mf.Type {
 	case dto.MetricType_GAUGE:
-		return makeSimpleDatapoints(*mf.Name, mf.Metric, sfxclient.GaugeF, gaugeExtractor)
+		makeGaugeDataPoints(sm, *mf.Name, mf.Metric, gaugeExtractor, applyDimensions)
 	case dto.MetricType_COUNTER:
-		return makeSimpleDatapoints(*mf.Name, mf.Metric, sfxclient.CumulativeF, counterExtractor)
+		makeSumDataPoints(sm, *mf.Name, mf.Metric, counterExtractor, applyDimensions)
 	case dto.MetricType_UNTYPED:
-		return makeSimpleDatapoints(*mf.Name, mf.Metric, sfxclient.GaugeF, untypedExtractor)
+		makeGaugeDataPoints(sm, *mf.Name, mf.Metric, untypedExtractor, applyDimensions)
 	case dto.MetricType_SUMMARY:
-		return makeSummaryDatapoints(*mf.Name, mf.Metric)
+		makeSummaryDatapoints(sm, *mf.Name, mf.Metric, applyDimensions)
 	// TODO: figure out how to best convert histograms, in particular the
 	// upper bound value
 	case dto.MetricType_HISTOGRAM:
-		return makeHistogramDatapoints(*mf.Name, mf.Metric)
+		makeHistogramDatapoints(sm, *mf.Name, mf.Metric, applyDimensions)
 	default:
-		return nil
 	}
 }
 
-func makeSimpleDatapoints(name string, ms []*dto.Metric, dpf dpFactory, e extractor) []*datapoint.Datapoint {
-	dps := make([]*datapoint.Datapoint, len(ms))
-	for i, m := range ms {
-		dps[i] = dpf(name, labelsToDims(m.Label), e(m))
+func makeGaugeDataPoints(sm pmetric.ScopeMetrics, name string, ms []*dto.Metric, e extractor, applyDimensions func(attrs pcommon.Map)) {
+	metric := sm.Metrics().AppendEmpty()
+	metric.SetName(name)
+	g := metric.SetEmptyGauge()
+	for _, m := range ms {
+		dp := g.DataPoints().AppendEmpty()
+		dp.SetDoubleValue(e(m))
+		for i := range m.Label {
+			dp.Attributes().PutStr(m.Label[i].GetName(), m.Label[i].GetValue())
+		}
+		applyDimensions(dp.Attributes())
 	}
-	return dps
 }
 
-func makeSummaryDatapoints(name string, ms []*dto.Metric) []*datapoint.Datapoint {
-	var dps []*datapoint.Datapoint
+func makeSumDataPoints(sm pmetric.ScopeMetrics, name string, ms []*dto.Metric, e extractor, applyDimensions func(attrs pcommon.Map)) {
+	metric := sm.Metrics().AppendEmpty()
+	metric.SetName(name)
+	sum := metric.SetEmptySum()
+	sum.SetAggregationTemporality(pmetric.AggregationTemporalityCumulative)
+	sum.SetIsMonotonic(true)
+	for _, m := range ms {
+		dp := sum.DataPoints().AppendEmpty()
+		dp.SetDoubleValue(e(m))
+		for i := range m.Label {
+			dp.Attributes().PutStr(m.Label[i].GetName(), m.Label[i].GetValue())
```

Comment:
You can get the attributes reference once and mutate instead for forming intermediate types w/ each call.